File: src/utils/exporter.ts
import fs from 'fs';
import { flattenObject, escapeCSVValue } from './cli-stream.js';

/**
 * Helper to flatten an EnrichedProfile for CSV/tabular output.
 * Mirroring the implementation used in CLI for consistency.
 */
function flattenEnriched(enriched: any): Record<string, any> {
  const base = flattenObject(enriched.profile ?? enriched);

  if (enriched.outcomes) {
    const o = enriched.outcomes;
    base['outcomes.credit.creditScore']            = o.credit?.creditScore;
    base['outcomes.credit.loanApprovalProbability']= o.credit?.loanApprovalProbability;
    base['outcomes.credit.approvedLoanAmountINR']  = o.credit?.approvedLoanAmountINR ?? '';
    base['outcomes.credit.reasonCodes']            = (o.credit?.reasonCodes ?? []).join('|');
    base['outcomes.health.healthRiskScore']        = o.health?.healthRiskScore;
    base['outcomes.health.bmiCategory']            = o.health?.bmiCategory;
    base['outcomes.health.healthcareAccessProbability'] = o.health?.healthcareAccessProbability;
    base['outcomes.health.likelyConditions']       = (o.health?.likelyConditions ?? []).join('|');
    base['outcomes.education.functionalLiteracy']  = o.education?.functionalLiteracy;
    base['outcomes.education.dropoutRisk']         = o.education?.dropoutRisk;
    base['outcomes.education.educationalMobility'] = o.education?.educationalMobility;
    base['outcomes.employment.employmentQuality']  = o.employment?.employmentQuality;
    base['outcomes.employment.expectedMonthlyWageINR'] = o.employment?.expectedMonthlyWageINR;
    base['outcomes.employment.wageGapRatio']       = o.employment?.wageGapRatio;
    base['outcomes.employment.vulnerabilityIndex'] = o.employment?.vulnerabilityIndex;
  }

  if (enriched.agentPersona) {
    base['persona.identityLine']   = enriched.agentPersona.identityLine;
    base['persona.worldview']      = enriched.agentPersona.beliefs?.worldview;
    base['persona.economicBehavior'] = enriched.agentPersona.economicBehavior;
    base['persona.stressResponse'] = enriched.agentPersona.stressResponse;
    base['persona.nationalPrevalence'] = enriched.agentPersona.nationalPrevalence;
  }

  if (enriched.narratives) {
    for (const doc of enriched.narratives) {
      base[`narrative.${doc.type}.wordCount`] = doc.metadata?.wordCount;
      base[`narrative.${doc.type}.language`]  = doc.language;
    }
  }

  return base;
}

/**
 * Formats a list of demographic profiles into a string representation of
 * JSON, JSONL, or CSV.
 * 
 * @param profiles - An array of standard DemographicProfile or EnrichedProfile objects.
 * @param format - Output format target ('json', 'jsonl', or 'csv').
 * @returns The formatted string dataset.
 */
export function formatProfiles(profiles: any[], format: 'json' | 'jsonl' | 'csv'): string {
  const fmt = format.toLowerCase();
  
  if (fmt === 'json') {
    return JSON.stringify(profiles, null, 2);
  }

  if (fmt === 'jsonl') {
    if (profiles.length === 0) return '';
    return profiles.map(p => JSON.stringify(p)).join('\n') + '\n';
  }

  if (fmt === 'csv') {
    if (profiles.length === 0) return '';
    
    // Check if profiles are enriched or standard
    const isEnriched = profiles[0].profile !== undefined || profiles[0].outcomes !== undefined;
    const flatProfiles = profiles.map(p => isEnriched ? flattenEnriched(p) : flattenObject(p));
    
    const headers = Object.keys(flatProfiles[0]);
    const headerRow = headers.map(escapeCSVValue).join(',');
    const dataRows = flatProfiles.map(flat => 
      headers.map(h => escapeCSVValue(flat[h])).join(',')
    );
    
    return [headerRow, ...dataRows].join('\n') + '\n';
  }

  throw new Error(`Unsupported format: ${format}. Supported formats are: json, jsonl, csv.`);
}

/**
 * Helper to save generated profiles directly to a file in standard format.
 * Automatically creates parent directories if they don't exist.
 * 
 * @param profiles - An array of standard DemographicProfile or EnrichedProfile objects.
 * @param filePath - Path to the file where data will be saved.
 * @param format - Output format target ('json', 'jsonl', or 'csv').
 */
export function saveProfilesToFile(profiles: any[], filePath: string, format: 'json' | 'jsonl' | 'csv'): void {
  const content = formatProfiles(profiles, format);
  fs.writeFileSync(filePath, content, 'utf8');
}

File: python/src/indian_fakedata/utils/exporter.py
"""
File Exporter Module

Provides zero-dependency JSON, JSONL, and CSV export for
generated demographic profiles and enriched profiles.
Uses only Python standard library modules.
"""

import json
import csv
import io


def _flatten_object(obj, prefix=""):
    """Recursively flatten a nested dict into a single-level key-value map."""
    result = {}
    if not obj or not isinstance(obj, dict):
        return result

    for key, value in obj.items():
        full_key = f"{prefix}_{key}" if prefix else key
        if isinstance(value, dict):
            result.update(_flatten_object(value, full_key))
        elif isinstance(value, list):
            if any(isinstance(v, dict) for v in value):
                result[full_key] = "; ".join(json.dumps(v) for v in value)
            else:
                result[full_key] = "; ".join(str(v) for v in value)
        else:
            result[full_key] = value

    return result


def _escape_csv_value(val):
    """Escape special characters for a CSV cell per RFC 4180."""
    if val is None:
        return ""
    if isinstance(val, dict) or isinstance(val, list):
        s = json.dumps(val)
    elif isinstance(val, bool):
        s = str(val)
    else:
        s = str(val)

    if ',' in s or '"' in s or '\n' in s or '\r' in s:
        return '"' + s.replace('"', '""') + '"'
    return s


def _flatten_enriched(enriched):
    """Flatten an EnrichedProfile for CSV output."""
    base = _flatten_object(enriched.get("profile", enriched))

    outcomes = enriched.get("outcomes")
    if outcomes:
        o = outcomes
        credit = o.get("credit", {})
        base["outcomes.credit.creditScore"] = credit.get("creditScore")
        base["outcomes.credit.loanApprovalProbability"] = credit.get("loanApprovalProbability")
        base["outcomes.credit.approvedLoanAmountINR"] = credit.get("approvedLoanAmountINR", "")
        base["outcomes.credit.reasonCodes"] = "|".join(credit.get("reasonCodes", []))

        health = o.get("health", {})
        base["outcomes.health.healthRiskScore"] = health.get("healthRiskScore")
        base["outcomes.health.bmiCategory"] = health.get("bmiCategory")
        base["outcomes.health.healthcareAccessProbability"] = health.get("healthcareAccessProbability")
        base["outcomes.health.likelyConditions"] = "|".join(health.get("likelyConditions", []))

        edu = o.get("education", {})
        base["outcomes.education.functionalLiteracy"] = edu.get("functionalLiteracy")
        base["outcomes.education.dropoutRisk"] = edu.get("dropoutRisk")
        base["outcomes.education.educationalMobility"] = edu.get("educationalMobility")

        emp = o.get("employment", {})
        base["outcomes.employment.employmentQuality"] = emp.get("employmentQuality")
        base["outcomes.employment.expectedMonthlyWageINR"] = emp.get("expectedMonthlyWageINR")
        base["outcomes.employment.wageGapRatio"] = emp.get("wageGapRatio")
        base["outcomes.employment.vulnerabilityIndex"] = emp.get("vulnerabilityIndex")

    persona = enriched.get("agentPersona")
    if persona:
        base["persona.identityLine"] = persona.get("identityLine")
        beliefs = persona.get("beliefs", {})
        base["persona.worldview"] = beliefs.get("worldview")
        base["persona.economicBehavior"] = persona.get("economicBehavior")
        base["persona.stressResponse"] = persona.get("stressResponse")
        base["persona.nationalPrevalence"] = persona.get("nationalPrevalence")

    narratives = enriched.get("narratives")
    if narratives:
        for doc in narratives:
            doc_type = doc.get("type", "unknown")
            metadata = doc.get("metadata", {})
            base[f"narrative.{doc_type}.wordCount"] = metadata.get("wordCount")
            base[f"narrative.{doc_type}.language"] = doc.get("language")

    return base


def format_profiles(profiles, fmt="json"):
    """
    Format a list of demographic profiles into a string.

    :param profiles: List of profile dicts (standard or enriched).
    :param fmt: Output format — 'json', 'jsonl', or 'csv'.
    :returns: Formatted string dataset.
    """
    fmt = fmt.lower()

    if fmt == "json":
        return json.dumps(profiles, indent=2, ensure_ascii=False)

    if fmt == "jsonl":
        if len(profiles) == 0:
            return ""
        return "\n".join(json.dumps(p, ensure_ascii=False) for p in profiles) + "\n"

    if fmt == "csv":
        if len(profiles) == 0:
            return ""

        is_enriched = "profile" in profiles[0] or "outcomes" in profiles[0]
        flat_profiles = [_flatten_enriched(p) if is_enriched else _flatten_object(p) for p in profiles]

        headers = list(flat_profiles[0].keys())
        header_row = ",".join(_escape_csv_value(h) for h in headers)
        data_rows = [
            ",".join(_escape_csv_value(flat.get(h)) for h in headers)
            for flat in flat_profiles
        ]
        return "\n".join([header_row] + data_rows) + "\n"

    raise ValueError(f"Unsupported format: {fmt}. Supported formats are: json, jsonl, csv.")


def save_profiles(profiles, filepath, fmt="json"):
    """
    Save generated profiles to a file.

    :param profiles: List of profile dicts (standard or enriched).
    :param filepath: Path to the output file.
    :param fmt: Output format — 'json', 'jsonl', or 'csv'.
    """
    content = format_profiles(profiles, fmt)
    with open(filepath, "w", encoding="utf-8") as f:
        f.write(content)
